File: src/py_pkg/py_pkg/topic_subscriber.py
import time

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class TopicSubscriber(Node):
    def __init__(self):
        super().__init__("py_topic_sub")
        self.subscription = self.create_subscription(
            String, "benchmark_topic", self.listener_callback, 10
        )
        self.get_logger().info("Python topic subscriber started")

    def listener_callback(self, msg):
        now_ns = time.time_ns()
        parts = msg.data.split("|", 1)
        if len(parts) == 2:
            sent_ns = int(parts[0])
            latency_us = (now_ns - sent_ns) / 1000.0
            self.get_logger().info(f"Received [{parts[1]}] latency={latency_us:.1f} us")


def main(args=None):
    rclpy.init(args=args)
    node = TopicSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: src/cpp_pkg/src/topic_subscriber.cpp
#include <chrono>
#include <string>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class TopicSubscriber : public rclcpp::Node {
public:
    TopicSubscriber() : Node("cpp_topic_sub") {
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "benchmark_topic", 10,
            std::bind(&TopicSubscriber::topic_callback, this, std::placeholders::_1));
        RCLCPP_INFO(this->get_logger(), "C++ topic subscriber started");
    }

private:
    void topic_callback(const std_msgs::msg::String::SharedPtr msg) {
        auto now = std::chrono::high_resolution_clock::now();
        auto now_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
            now.time_since_epoch()).count();
        auto sep = msg->data.find('|');
        if (sep != std::string::npos) {
            auto sent_ns = std::stoll(msg->data.substr(0, sep));
            auto latency_us = (now_ns - sent_ns) / 1000.0;
            RCLCPP_INFO(this->get_logger(), "Received [%s] latency=%.1f us",
                msg->data.substr(sep + 1).c_str(), latency_us);
        }
    }

    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TopicSubscriber>());
    rclcpp::shutdown();
    return 0;
}
